src/analysis.lisp

```diff
@@ -0,0 +1,100 @@
+;;;; Author: Mark Polyakov, released under MIT License
+
+(in-package #:cluck)
+
+(defclass e-analyzer ()
+  ()
+  (:documentation "Simple e-class analyses may be created by creating a subclass of E-CLASS-ANALYZER which implements E-CLASS-ANALYZER-ATOM and E-CLASS-ANALYZER-MERGE. More complex analyses may be implemented by subclassing E-CLASS directly."))
+
+(defclass e-analyzer-cost (e-analyzer)
+  ()
+  (:documentation "An e-class analyzer which can be used as the cost function when determining the best concrete representation of an e-class. Should implement E-CLASS-ANALYZER-COMPARE. A canned implementation of E-ANALYZER-MERGE exists for instances, which simply finds the cheapest e-node according to the analysis value; this implementation must not be overridden."))
+
+(defgeneric e-analyzer-e-node-value (analyzer e-node-car children)
+  (:documentation "Return the analysis value for an e-node. CHILDREN is a list of analysis values of the children. Some elements of CHILDREN may have circular references back to the e-class containing this e-node. In that case, E-ANALYZER-PARENT-VALUE is called to determine a reasonable default value. Note that in some situations, a circular reference will not use the parent value, depending on order of traversal, and instead will just be the value not including this new node."))
+(defgeneric e-analyzer-compare (analyzer v1 v2)
```

Comment:
ideally we put new lines between all definition forms. we can punt that until later.